File: PDS/DSP4/DSP4.m
% Juan Agustin Avila
% Julio 2020
% Reg 26076 - ELO
clc;clear all;
load('filtros.mat');    %se carga el archivo con los valores de los filtros
%% Punto 1

FiltrosFIR=[FiltroFIR_HP150Hz;FiltroFIR_BP40_100Hz;FiltroFIR_LP40Hz];

%con filterDesigner se exportaba el filtro IIR como una
%matriz SOS y un vector de ganancia G, eso se paso a una FT
[a,b]=sos2tf(SOSHP150Hz,GHP150Hz);
FiltroIIR_HP150Hz=[b a];        %Y luego numerador y denominador se unieron
[a,b]=sos2tf(SOSBP,GBP);
FiltroIIR_BP40_100Hz=[b a];
[a,b]=sos2tf(SOSLP40Hz,GLP40Hz);
FiltroIIR_LP40Hz=[b a];
FiltrosIIR=[FiltroIIR_HP150Hz;FiltroIIR_BP40_100Hz;FiltroIIR_LP40Hz];

nombre_senial=["OndaCuadrada","Perro"];
puntos=[512 4096];
offset=[10 500];
for j=1:length(nombre_senial)
    Nombres=["FiltroFIR_HP150Hz","FiltroFIR_BP40_100Hz","FiltroFIR_LP40Hz"];
    n=length(Nombres);
    for i=1:n
        nombre_filtro=Nombres(i);
        filtro=FiltrosFIR(i,:);
        filt(nombre_filtro,filtro,nombre_senial(j),puntos(j),offset(j));
    end
    
    %% filtros IIR
    Nombres=["FiltroIIR_HP150Hz","FiltroIIR_BP40_100Hz","FiltroIIR_LP40Hz"];
    for i=1:n
        nombre_filtro=Nombres(i);
        filtro=FiltrosIIR(i,:);
        filt(nombre_filtro,filtro,nombre_senial(j),puntos(j),offset(j));
    end
end

%% Graficacion en arduino:
figure();
for i=1:n
    nombre=Nombres(i);
    subplot((n*100)+10+i);
    graficacionfunciones(nombre,nombre_senial(1));
end


%% Realizacion del ecualizador
puntos=2048; ts=11025;
Gtotal=eye(5);  %Matriz identidad para probar cada filtro
for i=1:5
    ganancia=Gtotal(i,:);   %selecciona la columna I
    sim('ecualizador.slx');
    transformada=fft(Salida.Data,puntos);
    figure();
    plot((1:puntos),abs(transformada)); grid; xlim([1 puntos/2]);
    title("Respuesta frecuencial en la banda "+i);
end
puntos=4096;
ganancia=[1 1 1 1 1];       %Ganancias unitarias
sim('ecualizador.slx');
transformada=fft(Salida.Data,puntos);
figure();
plot((1:puntos),abs(transformada));
ganancia=[.001 4 8 2 .1]; %distintas ganancias por banda
sim('ecualizador.slx');
transformada=fft(Salida.Data,puntos);
hold on; grid on; 
plot((1:puntos),abs(transformada));xlim([1 puntos/2]);
legend("Señal con ganancias unitarias","Señal con distintas ganancias");
title("Comparacion entre señal original y con distintas ganancias");
%% funciones utilizadas
function filt(nombre_filtro,filtro,nombre_senial,puntos,offset)
    extra1="_Transformada";extra2="Puntos";%variables extra para nombres
    arch = fopen(nombre_filtro+".txt", 'wt');%abre un archivo con el nombre del filtro
    for i=length(filtro):-1:1               %y lo guarda de atras para adelante
        fprintf(arch, '%.10f\n', filtro(i));    %para facilitar el procesado en C
    end
    fclose(arch);
    system("DSP4.exe "+nombre_filtro+" "+nombre_senial);
    %ejecuta el programa en C con el nombre del filtro y el nombre de la señal
    original=load(nombre_senial+".txt");        %luego, carga ambos archivos
    filtrado=load(nombre_senial+nombre_filtro+".txt");  %el original y el filtrado
    figure();   %grafica ambas señales
    subplot(211);
    plot(original(2:length(original))); xlim([offset puntos]);grid on;
    title("Señal "+nombre_senial+" original");
    subplot(212);
    plot(filtrado(2:length(filtrado))); xlim([offset puntos]);grid on;
    title("Señal "+nombre_senial+" con el filtro "+nombre_filtro+" aplicado");
    %luego, realiza la transformada de ambas señales, la original y la filtrada
    system("..\DSP3\DSP3.exe "+nombre_senial+" "+puntos+" "+offset)
    system("..\DSP3\DSP3.exe "+nombre_senial+nombre_filtro+" "+puntos+" "+offset)
    %vuelve a cargarlas
    original=load(nombre_senial+extra1+puntos+extra2+".txt");
    filtrado=load(nombre_senial+nombre_filtro+extra1+puntos+extra2+".txt");
    figure();   %y grafica los resultados
    subplot(211);
    plot(original(2:length(original))); xlim([1 puntos]);grid on;
    title("Transformada de la señal "+nombre_senial+" con "+puntos+" puntos");
    subplot(212);
    plot(filtrado(2:length(filtrado)));  xlim([1 puntos]);grid on;
    title("Transformada de "+puntos+" puntos de la señal "+nombre_senial+" filtrada con "+nombre_filtro);

end

function graficacionfunciones(nombre,nombre_senial) %%abre el archivo y lo grafica
    arch=load(nombre_senial+nombre+"Arduino.txt");
    plot(arch(2:length(arch)));
    title("señal '"+nombre_senial+"' fitrada con el filtro "+nombre+" en arduino");   %Agrega titulo
    grid on;xlim([20 140]);
end

File: PDS/DSP4/DSP4.c
//Juan Agustin Avila
//julio 2020
//Reg 26076 - ELO - FI UNSJ

//TODO: revisar como pasarle argumentos con getopt

#include <unistd.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <stdlib.h>

FILE *abrir_archivo(char modo[], char nombre[], char extra[]) //abre los archivos en el modo indicado, se le puede agregar un sufijo al nombre
{
	FILE *archivo;
	char string[150];
	char cwd[PATH_MAX];
	getcwd(cwd, sizeof(cwd));											  //obtiene el directorio donde se esta ejecutando
	snprintf(string, sizeof(string), "%s\\%s%s.txt", cwd, nombre, extra); //siguen el patron original, solo les cambia el numero
	archivo = fopen(string, modo);
	if (archivo == NULL)
	{
		printf("No se pudo abrir el archivo \"%s%s.txt\"\n", nombre, extra); //no tiene mucho sentido porque no hace nada
	}
	return archivo;
}

int contar_lineas(FILE *archivo)
{
	int n = 0;
	float numero;
	while (!feof(archivo))
	{
		fscanf(archivo, "%f\n", &numero);
		n++;
	}
	rewind(archivo);
	return n;
}

float *carga_filtro(FILE *archfiltro, int N, int offset)
{
	int i = 0;
	float *filtro = malloc(sizeof(float) * (N - offset));
	for (i = 0; i < offset; i++)
	{
		fscanf(archfiltro, "%f\n", &filtro[0]);
	}

	for (i = 0; i < (N - offset); i++)
	{
		fscanf(archfiltro, "%f\n", &filtro[i]);
		printf("%f\n", filtro[i]);
	}
	rewind(archfiltro);
	return filtro;
}

void filtrado(FILE *archivo_original, float num[], float den[], int N, char nombre[], char extra[])
{
	float entrada[N], freq, out = 0, salida[N];
	int k = 0;
	FILE *archivo_nuevo;
	archivo_nuevo = abrir_archivo("w+", nombre, extra); //abre un nuevo archivo
	fscanf(archivo_original, "%f\n", &freq);			//obtiene la frecuencia
	fprintf(archivo_nuevo, "%.1f\n", freq);				//y la guarda en el nuevo
	for (k = 0; k < N; k++)								//inicializa el arreglo
	{
		entrada[k] = 0; //coloca todos los valores en cero
		salida[k] = 0;
	}

	while (!feof(archivo_original)) //lee una nueva linea
	{
		out = 0;
		for (k = 1; k < N; k++)
		{
			entrada[k - 1] = entrada[k]; //desplaza los valores
			salida[k - 1] = salida[k];
		}
		salida[N - 1] = 0;	//en principio se pone en cero para que no se calcule
		fscanf(archivo_original, "%f", &entrada[N - 1]);
		for (k = 0; k < N; k++)
		{
			out = out + (num[k] * entrada[k]) - (den[k] * salida[k]); //realiza el calculo
		}
		salida[N - 1] = out;	//finalmente se asigna el valor de salida actual al arreglo
		fprintf(archivo_nuevo, "%.3f\n", out);
	}

	fclose(archivo_nuevo); //cierra el archivo nuevo
	rewind(archivo_original);
}

int main(int argc, char *argv[])
{

	int N, i;
	FILE *arch;
	FILE *filtroarch;
	float *denominador;
	//TODO: reemplazar argc por getopt
	if (argc == 3)
	{
		printf("Se aplicara el filtro %s a la senial %s\n", argv[1], argv[2]);
	}
	else if (argc > 3)
	{
		printf("Demasiados argumentos.\n");
		return (1);
	}
	else
	{
		printf("No se pasaron suficientes argumentos.\n\n");
		printf("El programa se debe correr de la siguiente manera:\n");
		printf("DSP3.exe filtro_a_aplicar archivo_a_analizar \n");
		printf("Por ejemplo para el archivo \"Tono_50Hz.txt\" con el filtro \"FiltroFIR_LP50Hz.txt\":\n");
		printf("\"DSP3.exe FiltroFIR_LP50Hz Tono_50Hz\"\n");
		printf("(No se debe incluir la terminacion .txt de los archivos)\n\n\n");
		printf("IMPORTANTE: el archivo de filtros debe tener la siguiente forma:\n");
		printf("b(n-5)\nb(n-4)\n...\nb(n)\n\nY si es IIR luego \n\na(n-5)\n...\na(n)");

		return (1);
	}
	filtroarch = abrir_archivo("r", argv[1], ""); //abre archivo del filtro
	arch = abrir_archivo("r", argv[2], "");		  //abre archivo de la señal
	N = contar_lineas(filtroarch);	//cuenta los valores del filtro

	/*ATENCION: supone siempre que el filtro es de cuarto orden, por lo tanto
	el numerador siempre tendra 5 terminos. Si es un filtro FIR, sus denominadores
	siempre seran 0, si es IIR los siguientes 5 valores del archivo del filtro seran
	el denominador. Por lo tanto, el archivo siempre debe tener 5 o 10 valores.
	Cualquier archivo con valores distintos a esto casi con seguridad fallará*/

	printf("Numerador: \n");
	float *numerador = carga_filtro(filtroarch, 5, 0);
	if (N == 5)
	{
		printf("Filtro FIR, no tiene denominador\n");
		for (i = 0; i < N; i++)
		{
			denominador[i] = 0;
		}
	}
	else
	{
		printf("Denominador:\n");
		denominador = carga_filtro(filtroarch, N, 5);
	}
	fclose(filtroarch);
	filtrado(arch, numerador, denominador, 5, argv[2], argv[1]);
	fclose(arch);
	printf("Filtro aplicado con exito.\n");
}